TensorLib/Tensor.lean

```diff
@@ -0,0 +1,135 @@
+import Mathlib.Tactic
+namespace TensorLib
+section TensorData
+
+variable (a: Type)
+
+-- I tried `td_node (x: TensorData) (xs: List TensorData)` to remove the redundant empty node
+-- from the type. (We can already represent the 0 shape with `td_root []`.)
+-- This design works, but this way the examples are more uniform; it looked weird to
+-- syntactically preference the first of every node list. The code has some annoying special
+-- cases for the empty node now, though, so it may be worth revisiting this.
+inductive TensorData where
+| td_root (xs: List a)
+| td_node (xs: List TensorData)
+  abbrev Shape := List Nat
+
+-- TODO: Make tail-recursive, perhaps when Lean will eliminite tail calls in mutual-recursion
+def hasShape (x: TensorData a) (shape: Shape): Bool :=
```

Comment:
It may be best to define hasShape as a predicate, could be easier to reason about. Then we can still provide an implementation and how that they agree.

Reply:
I did this below as an inductive predicate `HasShape`. Was that not what you meant?

Reply:
I missed it.